File: client/src/app/models/basket.ts
import { Product } from "./product";

export type Basket = {
    basketId: string;
    items: Item[];
    clientSecret?: string;
    paymentIntentId?: string;
};

export class Item {
    constructor(product: Product, quantity: number) {
        this.productId = product.id;
        this.name = product.name;
        this.price = product.price;
        this.pictureUrl = product.pictureUrl;
        this.brand = product.brand;
        this.type = product.type;
        this.quantity = quantity;
    } // khi khỏi tạo class này và truyền đối tượng product và quantity thì output sẽ là object chứa các cặp key value được định nghĩa bên trong

    productId: number;
    name: string;
    price: number;
    pictureUrl: string;
    brand: string;
    type: string;
    quantity: number;
}

File: API/Entities/Basket.cs
using System;

namespace API.Entities;

public class Basket
{
    public int Id { get; set; }
    public required string BasketId { get; set; }
    public List<BasketItem> Items { get; set; } = []; //định nghĩa mỗi quan hệ với hảng basketItem
    public string? ClientSecret { get; set; }
    public string? PaymentIntentId { get; set; }

    public void AddItem(Product product, int quantity)
    {
        if (product == null) ArgumentNullException.ThrowIfNull(product);
        if (quantity < 0) throw new ArgumentException("Số lượng phải lớn hơn 0", nameof(quantity));

        var existingItem = FindItem(product.Id);

        if (existingItem == null)
        {
            Items.Add(new BasketItem
            {
                Product = product,
                Quantity = quantity
            });
        }
        else
        {
            existingItem.Quantity += quantity;
        }
    }

    public void RemoveItem(int productId, int quantity)
    {
        if (quantity <= 0) throw new ArgumentException("Số lượng phải lớn hơn 0", nameof(quantity));

        var item = FindItem(productId);

        if (item == null) return;

        item.Quantity -= quantity;

        if (item.Quantity <= 0) Items.Remove(item);
    }

    private BasketItem? FindItem(int productId) // trả về 1 basketitem hoặc ?(là null)
    {
        return Items.FirstOrDefault(item => item.ProductId == productId);
    }
}
